During the test, after the temporary port is closed, it may occasionally appear that the port is not completely released

**Describe the bug**
A clear and concise description of what the bug is.
```
[2022-12-14 07:49:23]--- FAIL: TestServerWithoutTlsEnabled (0.00s)
[2022-12-14 07:49:23]    tls_integration_test.go:92:
[2022-12-14 07:49:23]        	Error Trace:	xxx/pkg/util/tls/test/tls_integration_test.go:92
[2022-12-14 07:49:23]        	            				xxx/pkg/util/tls/test/tls_integration_test.go:178
[2022-12-14 07:49:23]        	Error:      	Received unexpected error:
[2022-12-14 07:49:23]        	            	listen tcp :41624: bind: address already in use
[2022-12-14 07:49:23]        	Test:       	TestServerWithoutTlsEnabled
[2022-12-14 07:49:23]time="2022-12-14T07:46:33+08:00" level=info msg="server listening on addresses" grpc="[::]:54039" http="[::]:43571"
[2022-12-14 07:49:23]2022/12/14 07:46:33 http: TLS handshake error from [::1]:44352: remote error: tls: bad certificate
[2022-12-14 07:49:23]2022/12/14 07:46:33 http: TLS handshake error from [::1]:44354: remote error: tls: bad certificate
[2022-12-14 07:49:23]time="2022-12-14T07:46:34+08:00" level=info msg="server listening on addresses" grpc="[::]:36626" http="[::]:49814"
[2022-12-14 07:49:23]2022/12/14 07:46:34 http: TLS handshake error from [::1]:34132: remote error: tls: bad certificate
[2022-12-14 07:49:23]2022/12/14 07:46:34 http: TLS handshake error from [::1]:34134: remote error: tls: bad certificate
[2022-12-14 07:49:23]2022/12/14 07:46:34 http: TLS handshake error from [::1]:34136: remote error: tls: bad certificate
[2022-12-14 07:49:23]time="2022-12-14T07:46:35+08:00" level=info msg="server listening on addresses" grpc="[::]:55998" http="[::]:46907"
[2022-12-14 07:49:23]2022/12/14 07:46:35 http: TLS handshake error from [::1]:55551: tls: client didn't provide a certificate
[2022-12-14 07:49:23]2022/12/14 07:46:35 http: TLS handshake error from [::1]:55553: tls: client didn't provide a certificate
[2022-12-14 07:49:23]2022/12/14 07:46:36 http: TLS handshake error from [::1]:55564: tls: client didn't provide a certificate
[2022-12-14 07:49:23]time="2022-12-14T07:46:36+08:00" level=info msg="server listening on addresses" grpc="[::]:39694" http="[::]:44458"
[2022-12-14 07:49:23]FAIL
[2022-12-14 07:49:23]FAIL	github.com/cortexproject/cortex/pkg/util/tls/test	4.287s
```

**To Reproduce**
Run：`make BUILD_IN_CONTAINER=false test`

**Expected behavior**
A clear and concise description of what you expected to happen.

**Environment:**
 - Infrastructure: [e.g., Kubernetes, bare-metal, laptop]
 - Deployment tool: [e.g., helm, jsonnet]

**Additional Context**
<!--  Additional relevant info which can help us debug this issue easily like Logs, Configuration etc. -->

--- a/pkg/util/tls/test/tls_integration_test.go
+++ b/pkg/util/tls/test/tls_integration_test.go
@@ -51,21 +51,21 @@ func (h *grpcHealthCheck) Watch(_ *grpc_health_v1.HealthCheckRequest, _ grpc_hea
 	return status.Error(codes.Unimplemented, "Watching is not supported")
 }
 
-func getLocalHostPort() (int, error) {
+func getLocalHostPort() (int, func() error, error) {
 	addr, err := net.ResolveTCPAddr("tcp", "localhost:0")
 	if err != nil {
-		return 0, err
+		return 0, nil, err
 	}
 
 	l, err := net.ListenTCP("tcp", addr)
 	if err != nil {
-		return 0, err
+		return 0, nil, err
 	}
 
-	if err := l.Close(); err != nil {
-		return 0, err
+	closePort := func() error {
+		return l.Close()
 	}
-	return l.Addr().(*net.TCPAddr).Port, nil
+	return l.Addr().(*net.TCPAddr).Port, closePort, nil
 }
 
 func newIntegrationClientServer(
@@ -80,9 +80,14 @@ func newIntegrationClientServer(
 		prometheus.DefaultRegisterer = savedRegistry
 	}()
 
-	grpcPort, err := getLocalHostPort()
+	grpcPort, closeGrpcPort, err := getLocalHostPort()
 	require.NoError(t, err)
-	httpPort, err := getLocalHostPort()
+	httpPort, closeHttpPort, err := getLocalHostPort()
+	require.NoError(t, err)
+
+	err = closeGrpcPort()
+	require.NoError(t, err)
+	err = closeHttpPort()
 	require.NoError(t, err)
 
 	cfg.HTTPListenPort = httpPort

--- a/pkg/util/tls/test/tls_integration_test.go
+++ b/pkg/util/tls/test/tls_integration_test.go
@@ -82,12 +82,12 @@ func newIntegrationClientServer(
 
 	grpcPort, closeGrpcPort, err := getLocalHostPort()
 	require.NoError(t, err)
-	httpPort, closeHttpPort, err := getLocalHostPort()
+	httpPort, closeHTTPPort, err := getLocalHostPort()
 	require.NoError(t, err)
 
 	err = closeGrpcPort()
 	require.NoError(t, err)
-	err = closeHttpPort()
+	err = closeHTTPPort()
 	require.NoError(t, err)
 
 	cfg.HTTPListenPort = httpPort

